### PR TITLE
Ucx+data handling

### PR DESCRIPTION
**Pull request by TomAugspurger**

### CODE REVIEW

File: distributed/comm/ucx.py [hunk 1]
```diff
@@ -316,4 +318,51 @@ def get_local_address_for(self, loc):
...
+        :func:`list.pop` to get the next receive.
+    gpu_inbound : int
+        The number of GPU recvies to do. Decrement this to get
+        back to regular recvs.
+
+    See Also
+    --------
+    should_peek : check whether it's appropriate to peek.
+    """
+    headers = msgpack.loads(frame, use_list=False)
```
Comment (mrocklin) [inline distributed/comm/ucx.py]: I recommend grepping for `msgpack.loads` to see how we usually do this.  You can avoid the bytestring keys if you want to with the `raw=False` keyword.

File: distributed/comm/ucx.py [hunk 2]
```diff
@@ -133,41 +133,42 @@ def peer_address(self) -> str:
...
+                this_size = sizes.pop()
+                # XXX: when do we get multiple keys here? Non-contiguous?
+                resp = await self.ep.recv_obj(this_size, cuda=bool(gpu_inbound))
+                # prepare for the next (header) recv
+                if gpu_inbound:
+                    gpu_inbound -= 1
+            else:
+                resp = await self.ep.recv_future()
             frame = ucp.get_obj_from_msg(resp)
+            if should_peek(frame):
```
Comment (mrocklin) [inline distributed/comm/ucx.py]: I wonder if we can maybe bake the gpu/cpu target information into an initial frame at the beginning of the message.

```diff
          frames = await to_frames(
              msg, serializers=serializers, on_error=on_error
          )
+         gpu_frame = b''.join([struct.pack(..., is_gpu_frame(frame) for frame in frames])
+         frames = [gpu_frame] + frames
          nframes = struct.pack("Q", len(frames))
```

File: distributed/protocol/cuda.py [hunk 3]
```diff
@@ -4,35 +4,36 @@
...
+    dtype = (0, x.dtype.str)
+
     header = x.__cuda_array_interface__.copy()
-    header['device'] = x.device.id
     header['lengths'] = (x.nbytes,)  # one per stride
     header['compression'] = (None,)  # TODO
-    # TODO: I don't think ucx-py should have to worry about
-    # MemoryPointer. Maybe some thin wrapper.
-    return header, [x.data]
+    header['is_cuda'] = 1
```
Comment (mrocklin) [inline distributed/protocol/cuda.py:27]: The approach suggested above would also stop the various serialization routines from having to know about device vs host frames, which will probably help future people who have to implement these.

File: distributed/comm/ucx.py (hunk 2 above)
Comment (TomAugspurger) [reply distributed/comm/ucx.py]: Hmm, that does sound much nicer. I'll take a look now.

**Comment by TomAugspurger** [comment]
@mrocklin 
https://github.com/TomAugspurger/distributed/pull/2/commits/8bc2dbb9ae495bc231d3d572ecdd89f7bb2c67ea implements your suggestion of included a couple header frames with descriptions of what's coming. I was able to remove all the peeking / deserialization code.

Do you have thoughts on whether it looks cleaner? It puts a bit tighter restriction on the "protocol" we're using to communication. Now the pattern is

1. message with the number of frames
2. message with whether each frame is a GPU frame
3. message with whether each frame is a sized frame
4. messages with all the "real" frames.

I may look into re-using `distributed.utils.nbytes`. I worry that the fallback of `sys.getsizeof(frame)` is not appropriate for this use case. We would rather fall back to `recv_future` if we aren't sure of the object's size.

**Comment by mrocklin** [comment]
I don't understand what a sized frame is.  We should know the size in bytes of every bytes-like object that we're sending through.  For Python objects like Pandas dataframes or cupy arrays there may be some ambiguity yes, but at the comms level we should be past all of that, and only hold onto bytes-like objects, for which I would expect the number of bytes of that object is very clear.  Is this not the case?

**Comment by TomAugspurger** [comment]
Hmm, that may be the case. Let me try it out on some workloads.

**Pull request merged**